File: 46-Permutations/46-Permutations.py
# Last updated: 2/12/2026, 3:45:40 PM
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        combinations = []
        used = [False] * len(nums)
        def recurse(rest : List[int]):
            if (len(rest) == len(nums)):
                combinations.append(rest[:])
                return
            i = 0
            while len(nums) > i:
                if (used[i]):
                    i += 1
                    continue
                rest.append(nums[i])
                used[i] = True
                recurse(rest)
                rest.pop()
                used[i] = False
                i += 1
        recurse([])
        return combinations

File: 46-Permutations/46-Permutations.js
// Last updated: 2/12/2026, 3:45:42 PM
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const totalCombinations = [];
    const used = new Array(nums.length).fill(false);
    function recurse(rest) {
        if (rest.length === nums.length) {
            totalCombinations.push([...rest]);
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            used[i] = true;
            rest.push(nums[i]);
            recurse(rest);
            rest.pop();
            used[i] = false;
        }
    }
    recurse([]);
    return totalCombinations;
};